### PR TITLE
Mod op results management

### PR DESCRIPTION
**Pull request by kazukiosawa**
## Main changes
- `asdfghjkl/vector.py`
  - Introduced a new `ParamVector` class that manages parameters (list of `torch.Tensor`) and vectors (dictionary which keys are the parameters and values are `torch.Tensor` with the same shapes as parameters) corresponding to them.
  - This class is used to represent matrix-vector products used in `fisher.py`, `hessian.py`, and `mvp.py`.
- `asdfghjkl/operations/operation.py`
  -  Made calculating layer-wise cov and cvp operations (`OP_COV`, `OP_CVP`).
  - Introduced a new `OperationManager` class
    - This class manages (register/call/clear) operations for modules.
    - The results created by operations can only be accessed through this class. 
    - Calculations of the full cov and cvp are done by this class. (`OP_FULL_COV`, `OP_FULL_CVP`)
- `asdfghjkl/core.py`
  - Under the `extend()` context, an instance of `OperationManager` is created to manage operations and results created by them. And the manager is yielded.
  - An instance of `ParamVector` can be passed to `extend()`. The vector is used by operations (i.e. calculating full/layer-wise cvp)
  - Added a new contextmanager `no_centered_cov()`, which wraps `extend()`, to be specialized for calculations in `fisher.py`
- `asdfghjkl/fisher.py`
  -  Modified to use an `OperationManager` (yielded by `with no_centered_cov() as cxt`) to calculate full/layer-wise fisher/fvp and access the operation results.
  - Removed `_register_fisher()`. Now the construction of the fisher (an `SymMatrix`) is done by `OperationManager`.
  - Simplifed `_accumulate_fisher()` and `_accumulate_fvp()`.
  - Modified to not use grad_scale for calculating fisher_exact and fisher_mc, instead, the scale is multiplied with the loss before doing a backpropagation on it.
  - Modified `fisher_free()` and `fisher_eig()` to use `ParamVector`.
- `asdfghjkl/mvp.py`
  - Modified `power_method()` and `conjugate_gradient_method()` to use `ParamVector`.
- `asdfghjkl/natural_gradient.py`
  - Modified `precondition()` and `precondition_module()` to use `ParamVector`.
  - Now any combination of fisher shapes is supported in `NaturalGradient`.
- `asdfghjkl/symmatrix.py`
  - Renamed `scaling` to `mul_` in `SymMatrix` class
  - `SymMatrix.mvp()` can receive `ParamVector` (for full matrix).
  - Modified `SymMatrix.mvp()`(for layer-wise matrix) and `Diag.mvp()` to allow module.weight with `requires_grad=False`

### CODE REVIEW

File: asdfghjkl/vector.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                self.vectors[p] = v
+        elif isinstance(vectors, dict):
+            for p in params:
+                assert p.shape == vectors[p].shape
+            self.vectors = vectors
+        else:
+            raise TypeError(f'Invalid vectors type: {type(vectors)}')
+
+    def values(self):
+        return self.vectors.values()
```
Comment (sxwang-oushu) [inline asdfghjkl/vector.py:39]: Here, `values()` should be guaranteed to have the same order as inserting. I think it might be wrong if python<3.6. https://stackoverflow.com/questions/34305003/difference-between-dictionary-and-ordereddict. So, personally, I prefer to use `OrderedDict`

File: asdfghjkl/vector.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        dist.all_reduce(packed_tensor)
+    else:
+        dist.reduce(packed_tensor, dst=0)
+    if all_reduce or is_master:
+        # unpack
+        rst = ParamVector(vectors.params, packed_tensor)
+    else:
+        rst = None
+
+    dist.barrier()
```
Comment (sxwang-oushu) [inline asdfghjkl/vector.py:115]: Does `all_reduce` have the implicit effect as `barrier`?

File: asdfghjkl/vector.py (hunk 1 above)
Comment (kazukiosawa) [reply asdfghjkl/vector.py:39]: Thanks, that's a good point! I will fix it to use `OrderedDict`

File: asdfghjkl/vector.py (hunk 2 above)
Comment (kazukiosawa) [reply asdfghjkl/vector.py:115]: It might have. But I think it is safe to put a barrier. This barrier is also for `reduce` (not only `all_reduce`).

File: asdfghjkl/vector.py (hunk 1 above)
Comment (kazukiosawa) [reply asdfghjkl/vector.py:38]: done!
